runtime/nvqir/photonics/PhotonicCircuitSimulator.h

```diff
@@ -0,0 +1,1217 @@
+    std::vector<std::size_t> tmp;                                              \
+    NAME(angle, tmp, quditIdxs);                                               \
+  }                                                                            \
+  virtual void NAME(const double angle,                                        \
+                    const std::vector<std::size_t> &controls,                  \
+                    const std::vector<std::size_t> quditIdxs) = 0;
+
+  /// @brief The plus gate
+  PHOTONIC_CIRCUIT_SIMULATOR_ONE_QUDIT(plus)
+  /// @brief The phase_shift gate
+  PHOTONIC_CIRCUIT_SIMULATOR_ONE_QUDIT_ONE_PARAM(phase_shift)
+  /// @brief The beam_splitter gate
+  PHOTONIC_CIRCUIT_SIMULATOR_TWO_QUDIT_ONE_PARAM(beam_splitter)
+
+// Undef those preprocessor defines.
+#undef PHOTONIC_CIRCUIT_SIMULATOR_ONE_QUDIT
+#undef PHOTONIC_CIRCUIT_SIMULATOR_ONE_QUDIT_ONE_PARAM
+#undef PHOTONIC_CIRCUIT_SIMULATOR_TWO_QUDIT_ONE_PARAM
+
+  /// @brief Measure the qudit with given index
+  virtual std::size_t mz(const std::size_t quditIdx) = 0;
+
+  /// @brief Measure operation. Here we check what the current execution
+  /// context is. If the context is sample, then we do nothing but store the
+  /// measure qudit, which we then use to do full state sampling when
+  /// flushAnySamplingTask() is called. If the context is sample-conditional,
+  /// then we have a circuit that contains if (`mz(q)`) and we measure the
+  /// qudit, collapse the state, and then store the sample qudit for final full
+  /// state sampling. We also return the bit result. If no execution context,
+  /// just measure, collapse, and return the bit.
+  virtual std::size_t mz(const std::size_t quditIdx,
+                         const std::string &registerName) = 0;
+
+  /// @brief Reset the qudit to the |0> state
+  virtual void resetQudit(const std::size_t quditIdx) = 0;
+
+  /// @brief Sample the current multi-qudit state on the given qudit indices
+  /// over a certain number of shots
+  virtual cudaq::ExecutionResult
+  sample(const std::vector<std::size_t> &quditIdxs, const int shots) = 0;
+
+  /// @brief Return the name of this PhotonicCircuitSimulator
+  virtual std::string name() const = 0;
+
+  /// @brief Return a thread_local pointer to this PhotonicCircuitSimulator
+  virtual PhotonicCircuitSimulator *clone() = 0;
+
+  /// Determine the (preferred) precision of the simulator.
+  virtual bool isSinglePrecision() const = 0;
+
+  bool isDoublePrecision() const { return !isSinglePrecision(); }
+
+}; // PhotonicCircuitSimulator
+
+/// @brief The PhotonicCircuitSimulatorBase is the type that is meant to
+/// be subclassed for new simulation strategies. The separation of
+/// PhotonicCircuitSimulator from PhotonicCircuitSimulatorBase allows simulation
+/// sub-types to specify the floating point precision for the simulation
+template <typename ScalarType>
+class PhotonicCircuitSimulatorBase : public PhotonicCircuitSimulator {
```

Comment:
Should the parent class be named with suffix 'Base'?

Reply:
Following the way CircuitSimulator and CircuitSimulatorBase  were implemented in `CircuitSimulator.h`
`class CircuitSimulatorBase : public CircuitSimulator`